Introduce function to copy chunk data between data nodes

pmwkaa opened the pull request:
Add internal copy_chunk_data() function which implements a way
to copy chunk data between data nodes using logical
replication.

Some moments to consider

* this functionality needs to be tested using separate instances,  `test_copy_chunk_data.sh` been introduced for that purpose
* no chunk locking been made during the copy operation
* no tests for corner cases such as connection failure during the copy or operation retry after failure
* there is no undo/clean procedure in case if operation fails during its operation

erimatnor commented on tsl/src/chunk_api.c:
```diff
@@ -1627,3 +1629,81 @@ chunk_api_get_chunk_relstats(PG_FUNCTION_ARGS)
 }
+
+Datum
+chunk_api_copy_data(PG_FUNCTION_ARGS)
```
Nit: Usually, `chunk_api` contains functions that are executed on the DNs, not the functions called from the AN.

erimatnor commented on tsl/src/chunk_api.c:
```diff
@@ -1627,3 +1629,81 @@ chunk_api_get_chunk_relstats(PG_FUNCTION_ARGS)
+													   NameStr(chunk->fd.table_name))),
+
+	/* create publication on the source data node */
+		ts_dist_cmd_run_on_data_nodes(psprintf("CREATE PUBLICATION %s FOR TABLE %s",
```
I am wondering if we should wrap these remote SQL commands in four API functions to be executed on the DNs:

```
chunk_create_subscription()
chunk_drop_subscription()
chunk_create_publication()
chunk_drop_publication()
```

Then a this  `copy_chunk` function is executed on the AN  and will call the above functions.

Why wrap in functions? I imagine that it makes testing easier, we can to parameter validation, and raise more appropriate errors for our use case.

Or is this over-doing it?

pmwkaa commented:
> Might make sense to provide sensible defaults for `retry_count` and `retry_delay_ms`

It is set to 1 hour by default. Do we want to make a macro from it?

nikkhils commented on sql/ddl_internal.sql:
```diff
@@ -37,3 +37,31 @@ CREATE OR REPLACE FUNCTION  _timescaledb_internal.chunk_drop_replica(
+        SELECT pgs.srsubstate = 'r'
+        INTO in_sync
+        FROM pg_subscription_rel pgs
+        JOIN pg_class pgc ON relname = table_name
```
We'd also need to use the `schema_name`.

afiskon commented on tsl/src/chunk.c:
```diff
@@ -421,3 +422,134 @@ chunk_drop_replica(PG_FUNCTION_ARGS)
+	NameData operation_id;
+	List *src_node;
+	List *dst_node;
+} ChunkCopyData;
```
I believe several comments to the structure and its fields would be appropriate.

afiskon commented on scripts/test_copy_chunk_data.sh:
```diff
@@ -0,0 +1,168 @@
+
+#
+# Test copy_chunk_data(): using logical replication to copy chunk data
+# between data nodes
```
Any chance we could utilize TAP tests in this one? I realize this is additional and probably unnecessary work and don't insist on it. On the other hand, Perl code is much easier to maintain than Bash and would be nice to be consistent with future tests.

afiskon commented on tsl/src/chunk.c:
```diff
@@ -421,3 +422,134 @@ chunk_drop_replica(PG_FUNCTION_ARGS)
+	 */
+	cmd = psprintf("SELECT pg_create_logical_replication_slot('%s', 'pgoutput')",
+				   NameStr(ccd->operation_id));
+	ts_dist_cmd_run_on_data_nodes(cmd, ccd->src_node, false);
```
Just curious, what will happen on data nodes when the access node fails after executing this code?

nikkhils replied:
The replication slot needs to be created on the destination node and not on the source node. I wonder how your test script is succeeding with this..

pmwkaa replied:
Previous effect of this command will still be remained and has to be manually cleaned up. We decided to address this issue separately

afiskon commented on tsl/src/remote/connection.c:
```diff
@@ -1390,6 +1408,60 @@ add_userinfo_to_server_options(ForeignServer *server, Oid user_id)
+	i = 0;
+	while (keywords[i] != NULL)
+	{
+		appendStringInfo(&buf, "%s=%s ", keywords[i], escape_param_str(values[i]));
```
Might be a good idea to call pfree() here on the string returned by escape_param_str()

svenklemm reviewed (changes requested):
The current implementation of the connection string handling allows for sql injection in connection string parameters. You shouldnt roll your own escape function. The correct escape for `'` is `''` unless you mark the string as having escape sequences.

nikkhils commented on .github/workflows/cron-tests.yaml:
```diff
@@ -81,3 +82,22 @@ jobs:
       run: |
         ./scripts/docker-run-restore-points-test.sh
+
+  copy_chunk_data:
```
This needs to be removed now that tap test infra is in place

nikkhils commented on tsl/src/chunk.c:
```diff
@@ -421,3 +422,134 @@ chunk_drop_replica(PG_FUNCTION_ARGS)
+	ts_dist_cmd_run_on_data_nodes(cmd, ccd->dst_node, false);
+
+	/* wait until data transfer finishes */
+	cmd = psprintf("CALL _timescaledb_internal.wait_subscription_sync(%s, %s)",
```
This will create an infinite loop wait because you are creating SUBSCRIPTION in the same transaction, so there won't be an entry that is visible in the `pg_subscription_rel` catalog yet. The wait has to happen after the CREATE SUBS/ALTER ENABLE has committed..

pmwkaa replied:
I guess we figured it out during discussions :)

pmwkaa commented:
Latest updates:

* Removed escape function and used `PQescapeString` analogue instead @svenklemm
* Removed bash/docker scripts to test copy chunk data and added similar TAP test instead @nikkhils, @afiskon

nikkhils commented on sql/ddl_internal.sql:
```diff
@@ -37,3 +37,33 @@ CREATE OR REPLACE FUNCTION  _timescaledb_internal.chunk_drop_replica(
+    schema_name    NAME,
+    table_name     NAME,
+    retry_count    INT DEFAULT 18000,
+    retry_delay_ms NUMERIC DEFAULT 0.200
```
`pg_sleep` takes seconds as input. So you are basically passing 0.2 seconds as default. Given than retry count is 10000, this will wait for a very long time.

pmwkaa replied:
This supposed to be an hour in total

svenklemm commented on tsl/src/remote/connection.c:
```diff
@@ -1468,6 +1486,45 @@ add_userinfo_to_server_options(ForeignServer *server, Oid user_id)
+		enlargeStringInfo(&buf, value_len * 2 + 1);
+
+		/* escape value */
+		buf.len += PQescapeString(buf.data + buf.len, values[i], value_len);
```
i think the correct function to use here is appendConnStrVal and then call PQescapeString on the whole string you pass as conninfo into CREATE SUBSCRIPTION, you are not treating the individual key-value pairs of the connection string correctly here, values have to be wrapped in quotes if they contain spaces.

pmwkaa replied:
It seems to be `appendConnStrVal` is part of the libpq source which is not part of imported PostgreSQL include directory we are using and sits in the `interfaces/libpq` directory. It is not clear how we should add this include path unless we always provide `PG_SOURCE_DIR` which we don't. Looks like it is easier to use `PQescapeString()` in both cases here

svenklemm commented on tsl/src/remote/connection.c:
```diff
@@ -1468,6 +1486,90 @@ add_userinfo_to_server_options(ForeignServer *server, Oid user_id)
+
+	initStringInfo(&connstr_escape);
+	enlargeStringInfo(&connstr_escape, connstr.len * 2 + 1);
+	connstr_escape.len += PQescapeString(connstr_escape.data, connstr.data, connstr.len);
```
i would prefer to see the sql escaping in the caller where you build the sql string, but other than that looks good
